File: intern/1510.stone-game-iv.js
/*
 * @lc app=leetcode id=1510 lang=javascript
 *
 * [1510] Stone Game IV
 */

// @lc code=start
/**
 * @param {number} n
 * @return {boolean}
 */
var winnerSquareGame = function (n) {
    /**
     * dp[i] 如果还剩下 i 个石子，能不能赢
     *
     * base case
     * dp[0] = false
     *
     * dp[i] =
     * for k from 1 to sqrt(i)
     * if !dp[i-k**2]
     *  dp[i] = true
     *  break;
     */

    const dp = new Array(n + 1);
    dp[0] = false;

    for (let i = 1; i <= n; i++) {
        dp[i] = false;
        const sqrt = Math.sqrt(i);
        for (let k = 1; k <= sqrt; k++) {
            if (!dp[i - k ** 2]) {
                // 对方会输
                dp[i] = true;
                break;
            }
        }
    }

    return dp[n];
};
// @lc code=end

File: src/2026/1510.stone-game-iv.cpp
/*
 * @lc app=leetcode id=1510 lang=cpp
 *
 * [1510] Stone Game IV
 */

#include <cmath>
#include <optional>
#include <vector>

// @lc code=start
class Solution {
 public:
  bool winnerSquareGame(const int n) {
    std::vector<std::optional<bool>> memo(n + 1);
    const bool can_win = CanWin(n, memo);
    return can_win;
  }

 private:
  static bool CanWin(const int n, std::vector<std::optional<bool>>& memo) {
    if (memo[n].has_value()) {
      return memo[n].value();
    }

    if (n == 0) {
      return false;
    }

    if (n == 1) {
      return true;
    }

    const int max_sqrt = static_cast<int>(std::sqrt(n));

    for (int sqrt = max_sqrt; sqrt >= 1; sqrt--) {
      const int current_pick = sqrt * sqrt;
      const bool opponent_can_win = CanWin(n - current_pick, memo);
      if (!opponent_can_win) {
        memo[n] = true;
        return true;
      }
    }

    memo[n] = false;
    return false;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.winnerSquareGame(2);
}
